docs: Access Monitoring Rules reference page (#55108)
gravitational/teleport: docs/pages/reference/access-controls/access-monitoring-rules.mdx

---
title: Access Monitoring Rule Reference
description: An explanation and overview of Access Monitoring Rules.
---

Access Monitoring Rules allow administrators to monitor Access Requests and apply
notification routing rules or automatic review rules based on specific conditions.

An Access Monitoring Rule can configure both a notification routing rule and an
automatic review rule, or it can contain only one.

## YAML specification

> 

Reviewer: I would add a brief introduction to the section here to save the reader a few cycles:

```suggestion

An Access Monitoring Rule is a dynamic Teleport resource with a structure similar to the following:

```

```yaml
kind: access_monitoring_rule
version: v1
metadata:
  name: example_rule
spec:
  # subjects specifies the kinds of subjects to monitor.
  # Possible values: "access_request"
  subjects:
  - access_request

  # condition specifies the conditions that should be met to apply the access
  # monitoring rule. The condition accepts a predicate expression which must
  # evaluate to a boolean value.
  #
  # This condition would be satisfied by an access request if the `access` role
  # is requested, and if the requesting user has the `team: dev` user trait.
  condition: |-

>     contains_all(set("access"), access_request.spec.roles) &&

Reviewer 2: `contains_all` with a set containing a single element is basically the same as `contains_any`, isn't it? Should we come up with a better example that illustrates `contains_all` or just change these both to `contains`?

    contains_any(user.traits["team"], set("dev"))

>   # Optional: desired_state specifies the desired reconciled state of the access
>   # request after the rule is applied. This field must be set to "reviewed" to

Reviewer: ```suggestion
  # Optional: desired_state specifies the desired reconciled state of the Access
  # Request after the rule is applied. This field must be set to "reviewed" to
```
For term consistency

    # integration specifies the name of an external integration source used to
    # help determine if a requesting user satisfies the rule conditions.
    # Use "builtin" to specify no external integration.
    # Possible values: "builtin"
    integration: builtin

    # decision determines whether to automatically approve or deny the
    # access request.
    # Possible values: "APPROVED" or "DENIED"
    decision: APPROVED

  # Optional: notification configures notification routing rules.
  notification:
    # name specifies the external integration to which the notifications should
    # be routed.
    # Possible values: "email", "discord", "slack", "pagerduty", "jira",
    # "mattermost", "msteams", "opsgenie", "servicenow", "datadog"
    name: email

    # recipients specifies the list of recipients to be notified when the
    # access monitoring rule is applied.
    recipients:
    - example@goteleport.com
```

## Notification routing rules

Administrators can use Access Monitoring Rules to route notifications to an external
notification system.

### Hosted integrations

To list the names of available hosted integrations you can use the following
command:
```code

> $ tctl get plugins --format=text

Reviewer 2: If `--format=text` is the default then I would omit it here.

Name          Status
------------- -----------
datadog       RUNNING
email         RUNNING
slack-default RUNNING
```

If no integrations are listed, enroll a plugin first.

> For a step by step guide on how to enroll a plugin see:
> [Access Request Plugins](../../admin-guides/access-controls/access-request-plugins/access-request-plugins.mdx)

Reviewer: ```suggestion
For a step by step guide on how to enroll a plugin see
[Access Request Plugins](../../admin-guides/access-controls/access-request-plugins/access-request-plugins.mdx).
```
Using a complete sentence

### Recipients

The accepted values for the `recipients` field depend on the selected
notification system.

| Notification System | Recipients                                                                            |
| ------------------- | ------------------------------------------------------------------------------------- |
| Datadog             | Email: `user@example.com`<br/>Team: `example-datadog-team`                            |
| Discord             | Channel ID: `...1234`                                                                 |
| Email               | Email: `user@example.com`                                                             |
| Jira                | -                                                                                     |
| Mattermost          | Email: `user@example.com`<br/>Team/Channel: `example-team/example-channel`            |

> | Microsoft Team      | Email: `user@example.com`<br/>Channel: `https://teams.microsoft.com/l/channel/...`    |

Reviewer 2: ```suggestion
| Microsoft Teams     | Email: `user@example.com`<br/>Channel: `https://teams.microsoft.com/l/channel/...`    |
```

| Opsgenie            | Schedule: `example-schedule`                                                          |
| PagerDuty           | Service: `example-service`                                                            |
| ServiceNow          | Schedule ID: `example-schedule-id`                                                    |
| Slack               | Email: `user@example.com`<br/>Channel: `example-channel`                              |

### Example notification routing rule

> If the following rule is configured, then anytime an Access Request for the

Reviewer: ```suggestion
If the following rule is configured, then any time an Access Request for the
```
Grammar fix

`access` role is created, a notification will be sent to the `#teleport-access`
Slack channel.

```yaml
kind: access_monitoring_rule
version: v1
metadata:
  name: slack-notifications
spec:
  subjects:
  - access_request
  condition: contains_any(access_request.spec.roles, set("access"))
  notification:
    name: slack-default
    recipients:
    - teleport-access
```

## Automatic review rules

Administrators can configure Access Monitoring Rules to automatically review an Access
Request when conditions are met.

To enable automatic reviews, set `spec.desired_state` to `reviewed` and define
`automatic_review`.

### Review decision

The `automatic_review.decision` option can be either `APPROVED` or `DENIED`.

> Notes:
> - Automatic reviews behave like regular reviews. They do not directly change the
>   state of the request and respect the required review thresholds.
> - If multiple automatic review rules match an Access Request, `DENIED` rules
>   take precedence.

Reviewer: ```suggestion
Automatic reviews behave like regular reviews. They do not directly change the state of the request and respect the required review thresholds.
  
If multiple automatic review rules match an Access Request, `DENIED` rules take precedence.
```

This is fine to treat as regular prose, I think, and doesn't need to be a list.

    integration: builtin
    decision: APPROVED
```

## Condition

The `condition` field is a predicate expression that evaluates to a boolean value
and determines which Access Requests the rule applies to.

Accepted fields within the condition predicate expression:
| Field                                   | Description                                    |
| --------------------------------------- | ---------------------------------------------- |
| access_request.spec.roles               | The set of roles requested.                    |
| access_request.spec.suggested_reviewers | The set of reviewers specified in the request. |
| access_request.spec.system_annotations  | A map of system annotations on the request.    |
| access_request.spec.user                | The requesting user.                           |
| access_request.spec.request_reason      | The request reason.                            |
| access_request.spec.creation_time       | The creation time of the request.              |
| access_request.spec.expiry              | The expiry time of the request.                |
| user.traits                             | A map of traits of the requesting user.        |

> Examples:

Reviewer: This looks a little cluttered to me: 
<img width="704" alt="image" src="https://github.com/user-attachments/assets/fb4c98a9-10d6-49b2-a955-b00c4e78fc23" />

I would use code fences for the example snippets make the text to the right of each dash either a comment above each snippet or regular text preceding each snippet.

## SSO users and IdP attributes

Access Monitoring Rules can be used with SSO users and attributes provided
by the IdP.

For example, if the following GitHub SSO configuration is used, GitHub users in
team `example-team` will be mapped to Teleport users with trait `github_teams: example-team`.

```yaml
# github.yaml
kind: github
version: v3
metadata:
  name: github
spec:
  teams_to_roles:
  - organization: example-org
    roles:
    - demo-access-request
    team: example-team
...
```

You can now create an Access Monitoring Rule that applies automatic reviews
based on the `github_teams` trait:

```yaml
kind: access_monitoring_rule
version: v1
metadata:
  name: dev-automatic-approval
spec:
  subjects:
  - access_request
  condition: |-
    contains_all(set("access"), access_request.spec.roles) &&
    contains_any(user.traits["github_teams"], set("example-team"))
  desired_state: reviewed
  automatic_review:
    integration: builtin
    decision: APPROVED
```

Trait mapping depends on the SSO provider. For configuration instructions, see:
[Configure Single Sign-On](../../admin-guides/access-controls/sso/sso.mdx)

> ## Access Monitoring Rule with IaC

Reviewer: ```suggestion
## Access Monitoring Rule with infrastructure as code
```
To make this explicit

> Access Monitoring Rules can be managed as infrastructure-as-code using Terraform.

Reviewer: ```suggestion
Access Monitoring Rules can be managed as infrastructure as code using Terraform.
```
No hyphens needed if this isn't a modifier
